// @locator/runtime/dist/components/Runtime.js
import { template as _$template } from "solid-js/web";
import { delegateEvents as _$delegateEvents } from "solid-js/web";
import { className as _$className } from "solid-js/web";
import { effect as _$effect } from "solid-js/web";
import { insert as _$insert } from "solid-js/web";
import { createComponent as _$createComponent } from "solid-js/web";
import { memo as _$memo } from "solid-js/web";
const _tmpl$ = /*#__PURE__*/_$template(`<div>`),
  _tmpl$2 = /*#__PURE__*/_$template(`<div class="fixed top-0 left-0 w-screen h-screen flex items-center justify-center bg-black/70 pointer-events-auto">`);
import { detectSvelte } from "@locator/shared";
import { batch, createEffect, createSignal, onCleanup, Show } from "solid-js";
import { render } from "solid-js/web";
import { isCombinationModifiersPressed } from "../functions/isCombinationModifiersPressed";
import { trackClickStats } from "../functions/trackClickStats";
import { MaybeOutline } from "./MaybeOutline";
import { SimpleNodeOutline } from "./SimpleNodeOutline";
import { IntroInfo } from "./IntroInfo";
import { Options } from "./Options";
import { bannerClasses } from "../functions/bannerClasses";
import BannerHeader from "./BannerHeader";
import { isExtension } from "../functions/isExtension";
import { NoLinkDialog } from "./NoLinkDialog";
import { WelcomeScreen } from "./WelcomeScreen";
import { isLocatorsOwnElement } from "../functions/isLocatorsOwnElement";
import { goToLinkProps } from "../functions/goTo";
import { getElementInfo } from "../adapters/getElementInfo";
import { getTree } from "../adapters/getTree";
import { TreeView } from "./TreeView";
import { OptionsProvider, useOptions } from "../functions/optionsStore";
import { DisableConfirmation } from "./DisableConfirmation";
function Runtime(props) {
  const [uiMode, setUiMode] = createSignal(["off"]);
  const [holdingModKey, setHoldingModKey] = createSignal(false);
  const [currentElement, setCurrentElement] = createSignal(null);
  const [dialog, setDialog] = createSignal(null);
  const [highlightedNode, setHighlightedNode] = createSignal(null);
  const options = useOptions();
  createEffect(() => {
    if (holdingModKey() && currentElement()) {
      document.body.classList.add("locatorjs-active-pointer");
    } else {
      document.body.classList.remove("locatorjs-active-pointer");
    }
  });
  function keyUpListener(e) {
    // if (e.code === "KeyO" && isCombinationModifiersPressed(e)) {
    //   if (uiMode()[0] === "tree") {
    //     setUiMode(["off"]);
    //   } else {
    //     setUiMode(["tree"]);
    //   }
    // }

    setHoldingModKey(isCombinationModifiersPressed(e));
  }
  function keyDownListener(e) {
    setHoldingModKey(isCombinationModifiersPressed(e));
  }
  function mouseOverListener(e) {
    setHoldingModKey(isCombinationModifiersPressed(e));
    const target = e.target;
    if (target && target instanceof HTMLElement) {
      // Ignore LocatorJS elements
      if (isLocatorsOwnElement(target)) {
        return;
      }
      batch(() => {
        setCurrentElement(target);
        // TODO: this is for highlighting elements in the tree, but need to move it to the adapter
        // if (solidMode()[0] === "tree" || solidMode()[0] === "treeFromElement") {
        //   const fiber = findFiberByHtmlElement(target, false);
        //   if (fiber) {
        //     const id = fiberToSimple(fiber, []);
        //     setHighlightedNode(id);
        //   }
        // }
      });

      // const found =
      //   target.closest("[data-locatorjs-id]") ||
      //   searchDevtoolsRenderersForClosestTarget(target);
      // if (found && found instanceof HTMLElement) {
      //   setCurrentElement(found);
      // }
    }
  }

  function mouseDownUpListener(e) {
    if (isCombinationModifiersPressed(e)) {
      e.preventDefault();
      e.stopPropagation();
    }
  }
  function clickListener(e) {
    if (!isCombinationModifiersPressed(e) && uiMode()[0] !== "options") {
      return;
    }
    const target = e.target;
    if (target && target instanceof HTMLElement) {
      if (target.shadowRoot) {
        return;
      }
      if (isLocatorsOwnElement(target)) {
        return;
      }
      const elInfo = getElementInfo(target, props.adapterId);
      if (elInfo) {
        const linkProps = elInfo.thisElement.link;
        if (linkProps) {
          e.preventDefault();
          e.stopPropagation();
          trackClickStats();
          if ((!isExtension() || detectSvelte()) && !options.getOptions().welcomeScreenDismissed) {
            setDialog(["choose-editor", linkProps]);
          } else {
            // const link = buidLink(linkProps, props.targets);
            goToLinkProps(linkProps, props.targets, options);
          }
        } else {
          console.error("[LocatorJS]: Could not find link: Element info: ", elInfo);
          setDialog(["no-link"]);
        }
      } else {
        console.error("[LocatorJS]: Could not find element info. Element: ", target);
        setDialog(["no-link"]);
      }
    }
  }
  function scrollListener() {
    setCurrentElement(null);
  }
  const roots = [document];
  document.querySelectorAll("*").forEach(node => {
    if (node.id === "locatorjs-wrapper") {
      return;
    }
    if (node.shadowRoot) {
      roots.push(node.shadowRoot);
    }
  });
  for (const root of roots) {
    root.addEventListener("mouseover", mouseOverListener, {
      capture: true
    });
    root.addEventListener("keydown", keyDownListener);
    root.addEventListener("keyup", keyUpListener);
    root.addEventListener("click", clickListener, {
      capture: true
    });
    root.addEventListener("mousedown", mouseDownUpListener, {
      capture: true
    });
    root.addEventListener("mouseup", mouseDownUpListener, {
      capture: true
    });
    root.addEventListener("scroll", scrollListener);
  }
  onCleanup(() => {
    for (const root of roots) {
      root.removeEventListener("keyup", keyUpListener);
      root.removeEventListener("keydown", keyDownListener);
      root.removeEventListener("mouseover", mouseOverListener, {
        capture: true
      });
      root.removeEventListener("click", clickListener, {
        capture: true
      });
      root.removeEventListener("mousedown", mouseDownUpListener, {
        capture: true
      });
      root.removeEventListener("mouseup", mouseDownUpListener, {
        capture: true
      });
      root.removeEventListener("scroll", scrollListener);
    }
  });
  function showTreeFromElement(element) {
    const newState = getTree(element);
    if (newState) {
      setUiMode(["tree", newState]);
    }
  }
  function openOptions() {
    setUiMode(["options"]);
  }
  return [_$memo((() => {
    const _c$ = _$memo(() => uiMode()[0] === "tree");
    return () => _c$() ? _$createComponent(TreeView, {
      get treeState() {
        return uiMode()[1];
      },
      close: () => setUiMode(["off"]),
      setTreeState: newState => setUiMode(["tree", newState]),
      get adapterId() {
        return props.adapterId;
      },
      get targets() {
        return props.targets;
      },
      setHighlightedNode: setHighlightedNode
    }) : null;
  })()), _$memo((() => {
    const _c$2 = _$memo(() => !!((holdingModKey() || uiMode()[0] === "options") && currentElement()));
    return () => _c$2() ? _$createComponent(MaybeOutline, {
      get currentElement() {
        return currentElement();
      },
      showTreeFromElement: showTreeFromElement,
      get adapterId() {
        return props.adapterId;
      },
      get targets() {
        return props.targets;
      }
    }) : null;
  })()), _$memo((() => {
    const _c$3 = _$memo(() => !!holdingModKey());
    return () => _c$3() ? (() => {
      const _el$ = _tmpl$();
      _$insert(_el$, _$createComponent(BannerHeader, {
        openOptions: openOptions,
        get adapter() {
          return props.adapterId;
        }
      }));
      _$effect(() => _$className(_el$, bannerClasses()));
      return _el$;
    })() : null;
  })()), _$memo((() => {
    const _c$4 = _$memo(() => !!highlightedNode());
    return () => _c$4() ? _$createComponent(SimpleNodeOutline, {
      get node() {
        return highlightedNode();
      }
    }) : null;
  })()), _$memo((() => {
    const _c$5 = _$memo(() => !!(!isExtension() && options.getOptions().showIntro !== false));
    return () => _c$5() ? _$createComponent(IntroInfo, {
      openOptions: openOptions,
      get hide() {
        return !!holdingModKey() || uiMode()[0] !== "off";
      },
      get adapter() {
        return props.adapterId;
      }
    }) : null;
  })()), _$memo((() => {
    const _c$6 = _$memo(() => uiMode()[0] === "options");
    return () => _c$6() ? _$createComponent(Options, {
      get adapterId() {
        return props.adapterId;
      },
      get targets() {
        return props.targets;
      },
      onClose: () => {
        setUiMode(["off"]);
      },
      showDisableDialog: () => {
        setUiMode(["disable-confirmation"]);
      },
      get currentElement() {
        return currentElement();
      }
    }) : null;
  })()), _$memo((() => {
    const _c$7 = _$memo(() => uiMode()[0] === "disable-confirmation");
    return () => _c$7() ? _$createComponent(DisableConfirmation, {
      onClose: () => {
        setUiMode(["off"]);
      }
    }) : null;
  })()), _$memo((() => {
    const _c$8 = _$memo(() => !!dialog());
    return () => _c$8() && (() => {
      const _el$2 = _tmpl$2();
      _el$2.$$click = e => {
        if (e.currentTarget === e.target) {
          setDialog(null);
        }
      };
      _$insert(_el$2, (() => {
        const _c$9 = _$memo(() => dialog()[0] === "no-link");
        return () => _c$9() && _$createComponent(NoLinkDialog, {});
      })(), null);
      _$insert(_el$2, (() => {
        const _c$10 = _$memo(() => dialog()[0] === "choose-editor");
        return () => _c$10() && _$createComponent(WelcomeScreen, {
          get targets() {
            return props.targets;
          },
          get originalLinkProps() {
            return dialog()[1];
          },
          onClose: () => {
            setDialog(null);
          }
        });
      })(), null);
      return _el$2;
    })();
  })())];
}
function RuntimeWrapper(props) {
  const options = useOptions();
  const isDisabled = () => options.getOptions().disabled || false;
  createEffect(() => {
    if (isDisabled() && isExtension()) {
      document.head.dataset.locatorDisabled = "disabled";
    } else {
      delete document.head.dataset.locatorDisabled;
    }
  });
  return _$createComponent(Show, {
    get when() {
      return !isDisabled();
    },
    get children() {
      return _$createComponent(Runtime, props);
    }
  });
}
export function initRender(solidLayer, adapter, targets) {
  render(() => _$createComponent(OptionsProvider, {
    get children() {
      return _$createComponent(RuntimeWrapper, {
        get targets() {
          return Object.fromEntries(Object.entries(targets).map(([key, t]) => {
            return [key, typeof t == "string" ? {
              url: t,
              label: key
            } : t];
          }));
        },
        adapterId: adapter
      });
    }
  }), solidLayer);
}
_$delegateEvents(["click"]);
